src/constraint.rs

```diff
@@ -608,7 +648,97 @@ impl FieldsConstraint {
 }
 
 impl ConstraintValidator for FieldsConstraint {
-    fn validate(&self, value: &OwnedElement, type_store: &TypeStore) -> ValidationResult {
-        todo!()
+    fn validate(
+        &self,
+        value: &OwnedElement,
+        type_store: &TypeStore,
+        open_open_content: bool,
+    ) -> ValidationResult {
+        let mut violations: Vec<Violation> = vec![];
+
+        // Check for null struct
+        if value.is_null() {
+            return Err(Violation::with_violations(
+                "fields",
+                ViolationCode::TypeMismatched,
+                &format!("Null struct not allowed for fields constraint"),
+                violations,
+            ));
+        }
+
+        // Create a peekable iterator for given struct
+        let ion_struct = match value.as_struct() {
+            None => {
+                return Err(Violation::with_violations(
+                    "fields",
+                    ViolationCode::TypeMismatched,
+                    &format!("expected struct ion found {}", value.ion_type()),
+                    violations,
+                ));
+            }
+            Some(ion_struct) => ion_struct,
+        };
+
+        // Verify if open content exists in the struct fields
+        if !open_open_content {
+            for (field_name, value) in ion_struct.iter() {
+                if !self.fields.contains_key(field_name.text().unwrap()) {
+                    violations.push(Violation::new(
+                        "fields",
+                        ViolationCode::InvalidOpenContent,
+                        &format!(
+                            "Found open open_content in the struct: {:?}: {:?}",
+                            field_name, value
+                        ),
+                    ));
+                }
+            }
+        }
+
+        // get the values corresponding to the field_name and perform occurs_validation based on the type_def
+        for (field_name, type_id) in &self.fields {
+            let type_def = type_store.get_type_by_id(*type_id).unwrap();
+            let values: Vec<&OwnedElement> = ion_struct.get_all(field_name).collect();
+
+            // perform occurs validation for type_def for all values of the given field_name
+            let occurs_range: Range = type_def
+                .get_occurs_constraint("fields")
+                .expect("Unable to parse occurs to range");
+
+            // verify if values follow occurs_range constraint
+            if !occurs_range
+                .contains(&(values.len() as i64).into())
+                .unwrap()
+            {
+                violations.push(Violation::new(
+                    "fields",
+                    ViolationCode::TypeMismatched,
+                    &format!(
+                        "Expected {:?} of type {:?}: found {}",
+                        occurs_range,
+                        type_def,
+                        values.len()
```

Comment:
minor issue—technically, we're checking the number of fields with a given name, not the number of elements with a type.
```suggestion
                        "Expected {:?} of field {:?}: found {}",
                        occurs_range,
                        field_name,
                        values.len()
```